core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterExchanger.java

```diff
@@ -0,0 +1,91 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+import io.trino.operator.exchange.PageReference.PageReleasedListener;
+import io.trino.spi.Page;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public class ScaleWriterExchanger
+        implements LocalExchanger
+{
+    private static final Logger log = Logger.get(ScaleWriterExchanger.class);
+
+    private final List<Consumer<PageReference>> buffers;
+    private final LocalExchangeMemoryManager memoryManager;
+    private final PageReleasedListener onPageReleased;
+    private final long bufferSizeLimit;
+    private final Supplier<Long> physicalWrittenBytesSupplier;
+    private final long writerMinSize;
+
+    // start with single writer and increase the writer count as buffer memory utilization increases
+    @GuardedBy("this")
+    private volatile int writerCount = 1;
+
+    @GuardedBy("this")
+    private volatile int nextWriterIndex = -1;
+
+    public ScaleWriterExchanger(
+            List<Consumer<PageReference>> buffers,
+            LocalExchangeMemoryManager memoryManager,
+            long maxBufferedBytes,
+            Supplier<Long> physicalWrittenBytesSupplier,
+            DataSize writerMinSize)
+    {
+        this.buffers = requireNonNull(buffers, "buffers is null");
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.onPageReleased = PageReleasedListener.forLocalExchangeMemoryManager(memoryManager);
+        this.bufferSizeLimit = maxBufferedBytes / 2;
+        this.physicalWrittenBytesSupplier = requireNonNull(physicalWrittenBytesSupplier, "physicalWrittenBytesSupplier is null");
+        this.writerMinSize = requireNonNull(writerMinSize, "writerMinSize is null").toBytes();
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Consumer<PageReference> buffer = buffers.get(getNextWriterIndex());
+        PageReference pageReference = new PageReference(page, 1, onPageReleased);
+        memoryManager.updateMemoryUsage(pageReference.getRetainedSizeInBytes());
+        buffer.accept(pageReference);
+    }
+
+    private synchronized int getNextWriterIndex()
+    {
+        if (writerCount < buffers.size()
+                && memoryManager.getBufferedBytes() >= bufferSizeLimit
```

Comment:
Calculate `memoryManager.getBufferedBytes()` and `physicalWrittenBytesSupplier.get()` before taking lock